### DISCUSSION TITLE
MQTT connection routine disconnects

### DISCUSSION THREAD

**Discussion by janossch** [Azure/azure-iot-sdk-java#1748] [Q&A] [answered]
Dear azure-iot-sdk-java developers,

We were excited to change to the latest library version specifically because of this change:
https://github.com/Azure/azure-iot-sdk-java/pull/1740

Before the change we saw warning level logs from Mqtt and IoTHubTransport classes.
However after to change to v2.3.0 we still see warning level logs from IoTHubTransport class (not from Mqtt).

We are creating our ModuleClient like below and opt out the log of routine disconnects.

```
ClientOptions options = ClientOptions.builder()
        .logRoutineDisconnectsAsErrors(false)
        .build();

ModuleClient.createFromEnvironment(unixDomainSocket, protocol, options);
```

Do you have any idea what could be missing?

Here is the log part which still appear, but I expected to be on information level and without the expection stack.

```
<4> 2023-10-16 12:36:14.949 +00:00 [WRN] - <IotHubTransport> <MQTT Rec: device/module> Updating transport status to new status DISCONNECTED_RETRYING with reason EXPIRED_SAS_TOKEN
com.microsoft.azure.sdk.iot.device.transport.ProtocolException: Mqtt connection lost
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:63)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connectionLost(Mqtt.java:335)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.connectionLost(CommsCallback.java:304)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:441)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:197)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: Connection lost (32109) - java.io.EOFException
	... 2 more
Caused by: java.io.EOFException
	at java.base/java.io.DataInputStream.readByte(Unknown Source)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:137)
	... 1 more
```

Thank you in advance!

**Comment by janossch** [comment]
It seems that the static method below overwrites the provided _ClientOptions_ if the _SSLContext_ not specified beforehand.

`ModuleClient.createFromEnvironment(unixDomainSocket, protocol, options);`

exact location:
https://github.com/Azure/azure-iot-sdk-java/blob/4ed458d7b03784478bd26290d407f0ada7591191/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java#L230C17-L230C31

I believe that line should consider the case if the _ClientOptions_ provided but without _SSLContext_.
@timtay-microsoft : what is your opinion?

**Reply by timtay-microsoft** [reply]
Looks like you are correct and this is a bug in the SDK. #1749 should fix this for you

**Reply by janossch** [reply]
Cool. Thank you! Allow me one more quick question related to this. When is the next release (v2.4.0) expected?

**Reply by timtay-microsoft** [reply]
I'll try to get a release out by the end of this week

**Reply by janossch** [reply]
@timtay-microsoft : Unfortunately we are still getting the _WARNING_ level logs from routine disconnect also with the new release v2.3.1.

Seems this line does not consider _ClientOptions_ provided by the user.
https://github.com/Azure/azure-iot-sdk-java/blob/da3fd071f23573c281e4df76f376a41ac35564d9/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/ModuleClient.java#L328

**Reply by timtay-microsoft** [reply, marked as answer]
#1757 should fix that

**Reply by janossch** [reply]
@timtay-microsoft sorry for the same question but when will the next release be expected which includes this fix?

**Reply by timtay-microsoft** [reply]
By the end of this week
